06_gpu_and_ml/embeddings/infinity/embedding_racetrack.py

```diff
@@ -0,0 +1,213 @@
+# * `max_concurrency` sets the [@modal.concurrent(max_inputs:int) ](https://modal.com/docs/guide/concurrent-inputs#input-concurrency "Modal: input concurrency") argument for the inference app. 
+# This takes advantage of the asynchronous nature of the Infinity embedding inference app.
+# * `gpu` is a string specifying the GPU to be used. 
+# * `max_containers` caps the number of containers allowed to spin-up.
+# * `max_ims` caps the number of images used in this example
+batch_size: int = 512
+max_concurrency: int = 2
+gpu: str = "H200"
+max_containers: int = 2
+max_ims: int = 20000
+imload_workers = os.cpu_count()  # Number of CPU thread workers used to load images when creating the batch
+
+# This should point to a model on huggingface that is supported by Infinity.
+# Note that your specifically chosen model might require specialized imports when
+# designing the image. This [OpenAI model](https://huggingface.co/openai/clip-vit-base-patch16 "OpenAI ViT")
+# takes about 4-10s to load into memory.
+MODEL_NAME = "openai/clip-vit-base-patch16"
+
+# ## Data setup
+# We use a [Modal Volume](https://modal.com/docs/guide/volumes#volumes "Modal.Volume") 
+# to store the images we want to encode.
+vol_name = "sweet-coral-db-20k"
+vol_mnt = Path("/data")
+vol = modal.Volume.from_name(vol_name)
+def find_images_to_encode(image_cap:int = 1) -> list[FileEntry]:
+    """
+    You can modify this function to find an return a list of your image paths.
+    """
+        
+    im_path_list = list(filter(lambda x: x.path.endswith(".jpg"),
+                         vol.listdir('/data', recursive=True)))
+    print(f"Found {len(im_path_list)} JPEGs, ", end='')
+
+    # Optional: cutoff number of images for testing (set to -1 to encode all)
+    if image_cap > 0:
+        im_path_list = im_path_list[:min(len(im_path_list), image_cap)]
+    return im_path_list
+
+# ## Define the image
+# Setup the image
+simple_image = (
+    modal.Image.debian_slim(python_version="3.10")
+    .pip_install(["infinity_emb[all]==0.0.76",   # for Infinity inference lib
+                  "sentencepiece",               # for this particular chosen model
+                  "more-itertools",              # for elegant list batching
+                  "torchvision"])                # for fast image loading
+    .env({"INFINITY_MODEL_ID": MODEL_NAME, "HF_HOME": "/data"})      # for Infinity inference lib and model caching
+)
+
+# Initialize the app
+app = modal.App(app_name, image=simple_image, volumes={vol_mnt: vol})
+
+# Imports inside the container
+with simple_image.imports():
+    from infinity_emb import AsyncEmbeddingEngine, EngineArgs
+    from infinity_emb.primitives import Dtype, InferenceEngine
+    from concurrent.futures import ThreadPoolExecutor
```

Comment:
Move 

```python
from concurrent.futures import ThreadPoolExecutor
```

to global imports. Let's leave only packages not available locally in the image imports.